### PR TITLE
More efficient SmartCacheLoader

### PR DESCRIPTION
**Pull request by cgrain**
### Description

On [this website](https://sagivtech.com/2017/09/19/optimizing-pytorch-training-code/) it is advised to delete tensors explicitly (with `del`) when the program is done with it. Applying that, I have made a small change in SmartCacheLoader that reflects that. I have also added a small unit test that ensures that the updated cache is indeed what it needs to be. 

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by rijobro** [comment]
Hi, I read through the blog you linked, and guess you're referring to this paragraph:

> Free up memory using del
This is a common pitfall for new PyTorch users, and we think it isn’t documented enough.
After you’re done with some PyTorch tensor or variable, delete it using the python del operator to free up memory.

In the blog post there is no data to back up this claim. As such, I think we'd need to see some numbers that your implementation uses less memory than the current one.

**Comment by rijobro** [comment]
Although I agree that the current implementation is a little ugly, it seems to have unnecessary for loops. How about:

```python
self._cache = self._cache[:remain_num] + self._replacements[:self._replace_num]
```

@wyli

**Comment by cgrain** [comment]
Thank you for your quick reply! 

Your question for some data is very reasonable, and I have only some anecdotal data unfortunately. I have actually implemented my method in my own project, and instead of SLURM killing my job because I was out of memory, the job ran without problems. Of course, I have not implemented nor tested your method. I will see if I can create a test that shows the difference in data use. The problem to check that thoroughly is that the garbage collector of python (which is not completely determenistic) can either be quick, which will result in a small difference, or slow, which will result in higher memory spikes.

**Comment by rijobro** [comment]
Sounds good, I'm curious to see what your benchmarking gives. My worry is that if we can't trust the garbage collector here, then we can't trust it anywhere in the code base and will have to delete everything everywhere. My default position is therefore to trust the garbage collector!

**Comment by wyli** [comment]
did a quick test it seems `test2` is good:
```py
import random

def test1(cache_num=100, replace_num=20):
  _cache = list(range(cache_num))
  replace = [random.random() for x in range(replace_num)]
  remain_num: int = cache_num - replace_num
  for i in range(remain_num):
    _cache[i] = _cache[i + replace_num]
  for i in range(replace_num):
    _cache[remain_num + i] = replace[i]
  return _cache

def test2(cache_num=100, replace_num=20):
  _cache = list(range(cache_num))
  replace = [random.random() for x in range(replace_num)]

  del _cache[:replace_num]
  _cache.extend(replace)
  return _cache

def test3(cache_num=100, replace_num=20):
  _cache = list(range(cache_num))
  replace = [random.random() for x in range(replace_num)]
  remain_num: int = cache_num - replace_num
  _cache = _cache[replace_num:replace_num + remain_num] + replace[:replace_num]
  return _cache
```

Screenshot 2021-05-12 at 13 45 25

**Comment by wyli** [comment]
thanks @cgrain, could you address this DCO issue by signing the commits?  https://github.com/Project-MONAI/MONAI/pull/2181/checks?check_run_id=2565463758

**Comment by cgrain** [comment]
Thanks @wyli, DCO should be signed now!

File: tests/test_smartcachedataset.py [hunk 1]
```diff
@@ -70,6 +71,35 @@ def test_shape(self, replace_rate, num_replace_workers, transform):
                         self.assertIsInstance(dataset[15]["image"], str)
                 dataset.shutdown()
 
+    def test_update_cache(self):
+        # Given
+        test_data = [{"image": f"test_image{i}.nii.gz", "label": f"test_image{i}.nii.gz"} for i in range(40)]
+        dataset = Smart_New_Dataset(
```
Comment (wyli) [inline tests/test_smartcachedataset.py:77]: Smart_New_Dataset is undefined here

**Review by wyli — APPROVED** [review]

**Pull request merged**